text/0000-expand-impl-trait.md

```diff
@@ -0,0 +1,853 @@
+  [associated type constructor](https://github.com/rust-lang/rfcs/pull/1598)
+  (ATCs), *not* just an associated type. That's one reason we didn't allow `impl
+  Trait` in traits in the first place. But we're now well on our way to
+  supporting ATCs in general, so extending the design is not so problematic.
+
+- The associated type (constructor) introduced by `impl Trait` in return
+  position is anonymous, and thus not usable within any function other than the
+  one that introduced it. In other words, the "scope" of the existential is just
+  that function. Larger scopes should be possible with the more expressive and
+  explicit syntax sketched in the appendix.
+
+### Functions in trait implementations
+
+When implementing a trait, `impl Trait` can -- and must -- be used *only* when
+it appears in the corresponding function signature in the trait's definition:
+
+```rust
+trait Foo {
+    fn bar(&self, f: impl FnOnce());
+    fn baz(&self) -> impl Iterator;
+}
+
+impl Foo for MyStruct {
+    fn bar(&self, f: impl FnOnce()) { ... }
+    fn baz(&self) -> impl Iterator { ... }
+}
+```
+
+The restrictions are largely there for simplicity; there are a few possible
+relaxations that a future RFC could tackle.
+
+### Associated types
+
+While existentials should eventually be usable when defining an associated type,
+this RFC proposes to defer that feature to a fully explicit syntax. The appendix
+at the end of the RFC sketches what that might look like.
+
+## Scoping for type and lifetime parameters
+
+In argument position, the type fulfilling an `impl Trait` is free to reference
+any types or lifetimes whatsoever. So in a signature like:
+
+```rust
+fn foo(iter: impl Iterator<Item = u32>);
+```
+
+the actual argument type may contain arbitrary lifetimes and mention arbitrary
+types. This follows from the desugaring to "anonymous" generic parameters.
+
+For return position, things are more nuanced.
+
+This RFC proposes that *all* type parameters are considered in scope for `impl
+Trait` in return position, as per Assumption 2 (which claims that this suffices
+for most use-cases) and Assumption 1 (which claims that we'll eventually provide
+an explicit syntax with finer-grained control).
+
+The lifetimes in scope include only those mentioned "explicitly" in a bound on
+the `impl Trait`. That is:
+
+- For `impl SomeTrait + 'a`, the `'a` is in scope for the concrete witness type.
+- For `impl SomeTrait + '_`, the lifetime that elision would imply is in scope
+  (this is again using the strawman shorthand syntax for an elided lifetime).
+
+Note, however, that the witness type can freely mention type parameters, which
+may themselves involve embedded lifetimes. Consider, for example:
+
+```rust
+fn transform(iter: impl Iterator<Item = u32>) -> impl Iterator<Item = u32>
+```
+
+Here, if the actual argument type was `SomeIter<'a>`, the return type can
+mention `SomeIter<'a>`, and therefore can *indirectly* mention `'a`.
```

Comment:
I'm a bit unclear on the rule for return types.

1. For the purpose of borrow checking, does an `impl Trait` return type *always* capture *all* input type parameters/input `impl Trait` types?

2. What about elided lifetime parameters? Is `fn foo(&self) -> impl Trait;` sugar for `fn foo<'a>(&'a self) -> impl Trait + 'a;`?

---

How about using the same lifetime inference rules as those used for references. That is, `&Type` is to `impl Trait` as `&'a Type` is to `impl Trait + 'a`. Under this rule,

```rust
 fn transform(&self, iter: impl Iterator<Item = u32>) -> impl Iterator<Item = u32>;
```

is just sugar for:

```rust
 fn transform<'a>(&'a self, iter: impl Iterator<Item = u32> + 'a) -> impl Iterator<Item = u32> + 'a;
```

To, e.g., exclude the `iter` parameter, one would use explicit lifetimes:

```rust
 fn transform<'a>(&'a self, iter: impl Iterator<Item = u32> /* + 'dontcare */) -> impl Iterator<Item = u32> + 'a;
```

(Or is this proposal equivalent to yours?)